src/serialization.jl

```diff
@@ -1,56 +1,87 @@
-const VERSION_KEY = "__versions__"
+const VERSIONS = Dict("Julia" => string(VERSION),
+                      "BenchmarkTools" => string(BENCHMARKTOOLS_VERSION))
 
-const VERSIONS = Dict("Julia" => string(VERSION), "BenchmarkTools" => string(BENCHMARKTOOLS_VERSION))
+# TODO: Add any new types as they're added
+const SUPPORTED_TYPES = [Benchmark, BenchmarkGroup, Parameters, TagFilter, Trial,
+                         TrialEstimate, TrialJudgement, TrialRatio]
 
-mutable struct ParametersPreV006
-    seconds::Float64
-    samples::Int
-    evals::Int
-    overhead::Int
-    gctrial::Bool
-    gcsample::Bool
-    time_tolerance::Float64
-    memory_tolerance::Float64
-end
-
-mutable struct TrialPreV006
-    params::Parameters
-    times::Vector{Int}
-    gctimes::Vector{Int}
-    memory::Int
-    allocs::Int
+for T in SUPPORTED_TYPES
+    @eval function JSON.lower(x::$T)
+        d = Dict{String,Any}()
+        for i = 1:nfields(x)
+            name = String(fieldname($T, i))
+            field = getfield(x, i)
+            value = typeof(field) in SUPPORTED_TYPES ? JSON.lower(field) : field
+            push!(d, name => value)
+        end
+        [string(typeof(x)), d]
+    end
 end
 
-function JLD.readas(p::ParametersPreV006)
-    return Parameters(p.seconds, p.samples, p.evals, Float64(p.overhead), p.gctrial,
-                      p.gcsample, p.time_tolerance, p.memory_tolerance)
+function recover(x::Vector)
+    length(x) == 2 || throw(ArgumentError("Expecting a vector of length 2"))
+    typename = x[1]::String
+    fields = x[2]::Dict
+    T = eval(parse(typename))::Type
```

Comment:
Isn't this a bit scary to call on something that comes from a json file? Should there be a whitelist of types and just switch on that?

Reply:
Yeah I suppose so. The hope is that people will only read in JSON files that were produced by this package, but I suppose it's pretty easy to falsify that if you know the expected format.

Reply:
Apparently Jarrett is unfazed by this.

Reply:
If it is valid JSON syntax, I guess it should be fine?